### ISSUE TITLE
saved parameters values don't match

### ISSUE THREAD

**Issue by michaelcapizzi** [clab/dynet#917]
I posted a question here: https://groups.google.com/forum/#!topic/dynet-users/uaJajWicH9g 

It's about this weird issue I'm having with my saved parameter values not matching those that I see when I call `dy.parameter(parameter_X).npvalue().

What's even weirder is when I built a simple, dummy example to accompany this issue, the problem did *not* occur.

Anyway, here's the toy example I built:

```
import dynet_config
import dynet as dy
import numpy as np

model = dy.ParameterCollection()

builders = [
    dy.GRUBuilder(1, 10, 7, model),
    dy.GRUBuilder(1, 10, 7, model)
]

pV = model.add_parameters((14, 2))
pb = model.add_parameters(2)

E = model.add_lookup_parameters((12, 8))
ET = model.add_lookup_parameters((7, 2))

trainer = dy.AdamTrainer(model)

dy.renew_cg()
# value of V before any training
V_original = dy.parameter(pV).npvalue()

# save without training
dy.save(
	"/home/mcapizzi/Desktop/test_model_without_training",
	[E, ET] + builders + [pV, pb]
)

# load and compare
loaded_model_without_training = dy.ParameterCollection()
loaded_params_without_training = dy.load(
	"/home/mcapizzi/Desktop/test_model_without_training",
	loaded_model_without_training
)

V_loaded_without_training = dy.parameter(loaded_params_without_training[4]).npvalue()

# compare V_original and V_loaded
print("V_original == V_loaded_without_training?")
print(np.all(V_original == V_loaded_without_training))

# train
for i in range(1000):
    dy.renew_cg()
    for datapoint, tags, labels in [
    ([1,2,3], [1, 1, 1], [0,0,1]),
    ([0,3,4], [2, 2, 2], [0,0,1])
    ]:
        V = dy.parameter(pV)
        b = dy.parameter(pb)
        words_in = [E[wid] for wid in datapoint]
        tags_in = [ET[tid] for tid in tags]
        init_states = builders[0].initial_state()
        init_states_b = builders[1].initial_state()
        in_ = [dy.concatenate([w, t]) for w,t in zip(words_in, tags_in)]
        in_b = [dy.concatenate([ww, tt]) for ww,tt in zip(reversed(words_in), reversed(tags_in))]
        hidden = [x.output() for x in init_states.add_inputs(in_)]
        hidden_b = [x.output() for x in init_states_b.add_inputs(in_b)]
        hidden_c = [dy.concatenate([h, hb]) for h, hb in zip(hidden, hidden_b)]
        outs = [dy.transpose(dy.transpose(hc) * V) + b for hc in hidden_c]
        losses = [dy.pickneglogsoftmax(o,l) for o,l in zip(outs, labels)]
        losses_summed = dy.esum(losses)
        losses_summed.backward()
        trainer.update()

dy.renew_cg()
V_after_training = dy.parameter(pV).npvalue()

print("testing")
dy.renew_cg()
for datapoint, tags, labels in [
    ([1,2,3], [1, 1, 1], [0,0,1]),
    ([0,3,4], [2, 2, 2], [0,0,1])
    ]:
    V = dy.parameter(pV)
    b = dy.parameter(pb)
    words_in = [E[wid] for wid in datapoint]
    tags_in = [ET[tid] for tid in tags]
    init_states = builders[0].initial_state()
    init_states_b = builders[1].initial_state()
    in_ = [dy.concatenate([w, t]) for w,t in zip(words_in, tags_in)]
    in_b = [dy.concatenate([ww, tt]) for ww,tt in zip(reversed(words_in), reversed(tags_in))]
    hidden = [x.output() for x in init_states.add_inputs(in_)]
    hidden_b = [x.output() for x in init_states_b.add_inputs(in_b)]
    hidden_c = [dy.concatenate([h, hb]) for h, hb in zip(hidden, hidden_b)]
    outs = [dy.transpose(dy.transpose(hc) * V) + b for hc in hidden_c]
    print(outs)


dy.renew_cg()
V_after_testing = dy.parameter(pV).npvalue()

dy.save(
    "/home/mcapizzi/Desktop/test_model_after_training",
	[E, ET] + builders + [pV, pb]
)

loaded_model_after_training = dy.ParameterCollection()

loaded_params_after_training = dy.load(
    "/home/mcapizzi/Desktop/test_model_after_training",
    loaded_model_after_training
)

# load V
dy.renew_cg()
V_loaded_after_training = dy.parameter(loaded_params_after_training[4]).npvalue()

print("V_loaded")
print(V_loaded_after_training)
print("V_after_testing")
print(V_after_testing)
print("V_after_training")
print(V_after_training)
```

But as I said, in this case the parameter values for `V` when loaded *do exactly* match those I see from `dy.parameter(pV).npvalue()`.  

My script that is having the problem is incredibly long and hard to follow, and so I do not want to subject you all to that.  But Graham suggested opening an issue nonetheless.

Thanks in advance for any advice, and for continuing to develop this great package.
-M

**Comment by chunyang-wen** [comment]
Cannot repeat the error with your code, so it is hard to debug.

**Comment by michaelcapizzi** [comment]
Thanks @chunyang-wen for having a look.  I understand that this is really weird, as the code snippet above works, but my (much more involved) code doesn't.

I'm just going crazy trying to figure out what's happening with my code.  But I will continue this conversation in the Google Group.  

Thanks again for your attempts to reproduce.

**Comment by neubig** [comment]
No, I'll keep this open at least until we have a chance to try it on bigger models and confirm that we're not able to reproduce.

**Comment by michaelcapizzi** [comment]
I was *finally* able to reproduce the issue I've been having.

Below is a simple `bi-directional RNN`.  I save the model in *two* ways before each datapoint:
1. using `dy.save()`
2. writing the `numpy array` to file

```
import dynet_config
dynet_config.set(random_seed=42)
import dynet as dy
import numpy as np

model = dy.ParameterCollection()

builders = [
    dy.GRUBuilder(1, 10, 7, model),
    dy.GRUBuilder(1, 10, 7, model)
]

pV = model.add_parameters((14, 2))
pb = model.add_parameters(2)

E = model.add_lookup_parameters((12, 8))
ET = model.add_lookup_parameters((7, 2))

trainer = dy.AdagradTrainer(model, learning_rate=0.2)
trainer.set_sparse_updates(False)

def run_lstm(datapoint, tags, labels):
    dy.renew_cg()
    V = dy.parameter(pV)
    b = dy.parameter(pb)
    words_in = [E[wid] for wid in datapoint]
    tags_in = [ET[tid] for tid in tags]
    init_states = builders[0].initial_state()
    init_states_b = builders[1].initial_state()
    in_ = [dy.concatenate([w, t]) for w,t in zip(words_in, tags_in)]
    in_b = [dy.concatenate([ww, tt]) for ww,tt in zip(reversed(words_in), reversed(tags_in))]
    hidden = [x.output() for x in init_states.add_inputs(in_)]
    hidden_b = [x.output() for x in init_states_b.add_inputs(in_b)]
    hidden_c = [dy.concatenate([h, hb]) for h, hb in zip(hidden, hidden_b)]
    outs = [dy.transpose(dy.transpose(hc) * V) + b for hc in hidden_c]
    losses = [dy.pickneglogsoftmax(o,l) for o,l in zip(outs, labels)]
    return losses

data = [
    ([1,2,3], [1, 1, 1], [0,0,1]),
    ([1,2,3], [1, 1, 1], [1,1,1])
]

# train
for i in range(3):
    for j in range(len(data)):
        datapoint = data[j][0]
        tags = data[j][1]
        labels = data[j][2]
        with open("/home/mcapizzi/Desktop/model_compare/test_model_{}_{}.txt".format(i, j), "w") as f:
            f.write(str(dy.parameter(pV).npvalue()) + "\n")
        dy.save(
            "/home/mcapizzi/Desktop/model_compare/test_model_{}_{}".format(i, j),
            [pV]
        )
        losses = run_lstm(datapoint, tags, labels)
        losses_summed = dy.esum(losses)
        losses_summed.backward()
        trainer.update()

with open("/home/mcapizzi/Desktop/model_compare/test_model_FINAL.txt", "w") as f:
    f.write(str(dy.parameter(pV).npvalue()) + "\n")
dy.save(
    "/home/mcapizzi/Desktop/model_compare/test_model_FINAL",
    [pV]
)
```

Before any training is done (`epoch=0, datapoint=0`), the values from `dy.save()` and those from `numpy array` match:
(**NOTE** The use of `random seed` should help for reproducibility of parameter values)

```
(sentence-segmentation) mcapizzi@nextiva-desktop:/media/mcapizzi/data/Github/sentence-segmentation/src/nn_scripts$ head ~/Desktop/model_compare/test_model_0_0.txt 
[[ 0.10505317  0.54068172]
 [ 0.16705965  0.15424396]
 [ 0.53916979 -0.13949521]
 [ 0.36858609  0.39244112]
 [ 0.09243665  0.56484079]
 [ 0.21698606  0.18553023]
 [-0.13696332  0.49645111]
 [ 0.08985593 -0.35923684]
 [ 0.17549148 -0.37257826]
 [-0.45499232 -0.27683997]

(sentence-segmentation) mcapizzi@nextiva-desktop:/media/mcapizzi/data/Github/sentence-segmentation/src/nn_scripts$ cat ~/Desktop/model_compare/test_model_0_0.data
#Parameter# /_0 {14,2} 449 ZERO_GRAD
+1.05053172e-01 +1.67059645e-01 +5.39169788e-01 +3.68586093e-01 +9.24366489e-02 +2.16986060e-01 -1.36963323e-01 +8.98559317e-02 +1.75491482e-01 -4.54992324e-01 -5.11295758e-02 +3.81145746e-01 +5.58689497e-02 +3.92701596e-01 +5.40681720e-01 +1.54243961e-01 -1.39495209e-01 +3.92441124e-01 +5.64840794e-01 +1.85530230e-01 +4.96451110e-01 -3.59236836e-01 -3.72578263e-01 -2.76839972e-01 -5.27422547e-01 -3.49555522e-01 -4.88945097e-01 -1.50326326e-01 
```

But when I inspect the parameter values for `epoch=1, datapoint=1` (which is after one `update`), they already differ:

```
(sentence-segmentation) mcapizzi@nextiva-desktop:/media/mcapizzi/data/Github/sentence-segmentation/src/nn_scripts$ head ~/Desktop/model_compare/test_model_0_1txt 
[[ 0.10505317  0.54068172]
 [ 0.16705965  0.15424396]
 [ 0.53916979 -0.13949521]
 [ 0.36858609  0.39244112]
 [ 0.09243665  0.56484079]
 [ 0.21698606  0.18553023]
 [-0.13696332  0.49645111]
 [ 0.08985593 -0.35923684]
 [ 0.17549148 -0.37257826]
 [-0.45499232 -0.27683997]

(sentence-segmentation) mcapizzi@nextiva-desktop:/media/mcapizzi/data/Github/sentence-segmentation/src/nn_scripts$ cat ~/Desktop/model_compare/test_model_0_1.data
#Parameter# /_0 {14,2} 449 ZERO_GRAD
-9.49468315e-02 -3.29403467e-02 +3.39169770e-01 +5.68586111e-01 +2.92436630e-01 +4.16986048e-01 +6.30366802e-02 +2.89855927e-01 +3.75491500e-01 -2.54992306e-01 -2.51129568e-01 +1.81145713e-01 +2.55868942e-01 +1.92701563e-01 +7.40681708e-01 +3.54243964e-01 +6.05048165e-02 +1.92441091e-01 +3.64840776e-01 -1.44697726e-02 +2.96451092e-01 -5.59236825e-01 -5.72578251e-01 -4.76839960e-01 -3.27422559e-01 -1.49555519e-01 -6.88945115e-01 +4.96736765e-02 
```

It appears as if `dy.parameter(pV).npvalue()` has *not* updated at all, while the output of `dy.save()` has updated.

And then, not surprising, the values of the parameters still differ after training is complete:

```
(sentence-segmentation) mcapizzi@nextiva-desktop:/media/mcapizzi/data/Github/sentence-segmentation/src/nn_scripts$ head ~/Desktop/model_compare/test_model_FINAL.txt 
[[ 0.10789666  0.53783816]
 [-0.08437252  0.4056761 ]
 [ 0.32101145  0.07866311]
 [ 0.38045615  0.3805711 ]
 [ 0.30201119  0.35526624]
 [ 0.43488705 -0.03237073]
 [ 0.14682768  0.21266009]
 [ 0.1616486  -0.43102953]
 [ 0.1121488  -0.30923557]
 [-0.2478199  -0.4840124 ]
(sentence-segmentation) mcapizzi@nextiva-desktop:/media/mcapizzi/data/Github/sentence-segmentation/src/nn_scripts$ cat ~/Desktop/model_compare/test_model_FINAL.data
#Parameter# /_0 {14,2} 449 ZERO_GRAD
+1.64894119e-01 -1.63068652e-01 +3.68678838e-01 +3.54553968e-01 +3.30585957e-01 +4.90389526e-01 +1.62173867e-01 +1.07020408e-01 +9.14504156e-02 -1.90508053e-01 -3.85245532e-01 +3.49520892e-02 +4.50982887e-04 +7.42758438e-02 +4.80840713e-01 +4.84372199e-01 +3.09956968e-02 +4.06473309e-01 +3.26691478e-01 -8.78731981e-02 +1.97313905e-01 -3.76401335e-01 -2.88537174e-01 -5.41324258e-01 -1.93306595e-01 -3.36188311e-03 -4.33527201e-01 +1.68099359e-01 
```

I stumbled across the issue because my training pipeline does the following:

1. train
2. save parameters
3. test
4. record results

But when I loaded parameters to test my `predict()` code, I noticed that, not surprisingly, the results differed when testing on the `loaded` parameters v. the parameters in memory from training.

Is there something wrong with my pipeline?
Should I *not* be testing the model in the same "session" as training?

**Comment by michaelcapizzi** [comment]
I finally figured out the issue!

If you recall, I had a situation where the parameter values saved did *not* equal those I saw when I called `dy.parameter(parameter_X).npvalue()`.

And I finally found that it has to do with the argument, `dynet_config.set(weight_decay).

See the sample script below:

```
import dynet_config
dynet_config.set(
     random_seed=42, 
     weight_decay=0.000001
)
import dynet as dy
import numpy as np

model = dy.ParameterCollection()

builders = [
    dy.GRUBuilder(1, 10, 7, model),
    dy.GRUBuilder(1, 10, 7, model)
]

pV = model.add_parameters((14, 2))
pb = model.add_parameters(2)

E = model.add_lookup_parameters((12, 8))
ET = model.add_lookup_parameters((7, 2))

trainer = dy.AdagradTrainer(model, learning_rate=0.2)    
trainer.set_sparse_updates(False)


def run_lstm(datapoint, tags, labels):
    V = dy.parameter(pV)
    b = dy.parameter(pb)
    words_in = [E[wid] for wid in datapoint]
    tags_in = [ET[tid] for tid in tags]
    init_states = builders[0].initial_state()
    init_states_b = builders[1].initial_state()
    in_ = [dy.concatenate([w, t]) for w,t in zip(words_in, tags_in)]
    in_b = [dy.concatenate([ww, tt]) for ww,tt in zip(reversed(words_in), reversed(tags_in))]
    hidden = [x.output() for x in init_states.add_inputs(in_)]
    hidden_b = [x.output() for x in init_states_b.add_inputs(in_b)]
    hidden_c = [dy.concatenate([h, hb]) for h, hb in zip(hidden, hidden_b)]
    outs = [dy.transpose(dy.transpose(hc) * V) + b for hc in hidden_c]
    losses = [dy.pickneglogsoftmax(o,l) for o,l in zip(outs, labels)]
    return losses


data = [
    ([1,2,3], [1, 1, 1], [0,0,1]),
    ([1,2,3], [1, 1, 1], [1,1,1])
]

# train
for i in range(3):
    for j in range(len(data)):
        datapoint = data[j][0]
        tags = data[j][1]
        labels = data[j][2]
        losses = run_lstm(datapoint, tags, labels)
        losses_summed = dy.esum(losses)
        losses_summed.backward()
        trainer.update()

dy.renew_cg()
with open("/home/mcapizzi/Desktop/model_compare/test_model_FINAL.txt", "w") as f:
    f.write(np.array_str(dy.parameter(pV).npvalue()) + "\n")
dy.save(
    "/home/mcapizzi/Desktop/model_compare/test_model_FINAL",
    [pV]
)
```
Here are the outputs of the two different sets of saved parameters: 

```
(sentence-segmentation) mcapizzi@nextiva-desktop:/media/mcapizzi/data/Github/sentence-segmentation/src/nn_scripts$ head ~/Desktop/model_compare/test_model_FINAL.txt
[[  1.64893270e-01   4.80837733e-01]
 [ -1.63067669e-01   4.84369308e-01]
 [  3.68677557e-01   3.09946593e-02]
 [  3.54551584e-01   4.06470954e-01]
 [  3.30583751e-01   3.26689690e-01]
 [  4.90387022e-01  -8.78731906e-02]
 [  1.62172750e-01   1.97312817e-01]
 [  1.07019641e-01  -3.76398921e-01]
 [  9.14504603e-02  -2.88535982e-01]
 [ -1.90506294e-01  -5.41321576e-01]
(sentence-segmentation) mcapizzi@nextiva-desktop:/media/mcapizzi/data/Github/sentence-segmentation/src/nn_scripts$ cat ~/Desktop/model_compare/test_model_FINAL.data
#Parameter# /_0 {14,2} 449 ZERO_GRAD
+1.64894268e-01 -1.63068667e-01 +3.68679792e-01 +3.54553729e-01 +3.30585748e-01 +4.90390003e-01 +1.62173733e-01 +1.07020289e-01 +9.14510190e-02 -1.90507457e-01 -3.85245681e-01 +3.49528976e-02 +4.50871274e-04 +7.42755979e-02 +4.80840653e-01 +4.84372258e-01 +3.09948474e-02 +4.06473428e-01 +3.26691687e-01 -8.78737271e-02 +1.97314024e-01 -3.76401216e-01 -2.88537741e-01 -5.41324854e-01 -1.93306491e-01 -3.36269569e-03 -4.33526993e-01 +1.68099612e-01 
```

I have since noticed this paragraph in the documentation (http://dynet.readthedocs.io/en/latest/unorthodox.html#weight-decay):
```
As described in the Command Line Options, weight decay is implemented through the option --dynet-weight-decay. If this value is set to wd, each parameter in the model is multiplied by (1-wd) after every parameter update. This weight decay is similar to L2 regularization, and is equivalent in the case of using simple SGD (SimpleSGDTrainer), but it is not the same when using any other optimizers such as AdagradTrainer or AdamTrainer. You can still try to use weight decay with these optimizers, and it might work, but if you really want to correctly apply L2 regularization with these optimizers, you will have to directly calculate the L2 norm of each of the parameters and add it to the objective function before performing your update.
```
I don't pretend to understand all the different `optimizers`, but I also tried the `SimpleSGDTrainer` and the same problem occurred.

**Comment by chunyang-wen** [comment]
Do you happen to use GPU ?

**Comment by michaelcapizzi** [comment]
Thank you @chunyang-wen .  I was not aware of the `.as_array()` function.  That is exactly what I have been looking for as I found `dy.parameter()` to be unnecessary.

**Comment by michaelcapizzi** [comment]
I'm happy to have this issue closed but will leave it to @neubig to make that decision, as he initially wanted to keep it open.

Thanks again for continuing to maintain this great package!

**Comment by neubig** [comment]
Looks resolved, thanks!
